Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderParser.cpp

```diff
@@ -0,0 +1,171 @@
+        "video/mp4; "
+        "audio/mp4; "
+        "audio/mpeg; "
+        "audio/x-flac; "
+        "audio/x-eac3; "
+        "audio/x-ac3; "
+        "video/x-h264; "
+        "video/x-h265; "
+        "video/x-vp9; video/x-vp8; "
+        "video/x-av1; "
+        "audio/x-opus; audio/x-vorbis"));
+
+static GRefPtr<GstCaps> createThunderParseSinkPadTemplateCaps()
+{
+    GRefPtr<GstCaps> caps = adoptGRef(gst_caps_new_empty());
+
+    auto& supportedKeySystems = CDMFactoryThunder::singleton().supportedKeySystems();
+
+    if (supportedKeySystems.isEmpty()) {
+        GST_WARNING("no supported key systems in Thunder, we won't be able to decrypt anything with the decryptor");
+        return caps;
+    }
+
+    for (const auto& mediaType : GStreamerEMEUtilities::s_cencEncryptionMediaTypes) {
+        gst_caps_append_structure(caps.get(), gst_structure_new_empty(mediaType.characters()));
+        gst_caps_append_structure(caps.get(), gst_structure_new("application/x-cenc", "original-media-type", G_TYPE_STRING, mediaType.characters(), nullptr));
+    }
+    for (const auto& keySystem : supportedKeySystems) {
+        for (const auto& mediaType : GStreamerEMEUtilities::s_cencEncryptionMediaTypes) {
+            gst_caps_append_structure(caps.get(), gst_structure_new("application/x-cenc", "original-media-type", G_TYPE_STRING,
+                mediaType.characters(), "protection-system", G_TYPE_STRING, GStreamerEMEUtilities::keySystemToUuid(keySystem), nullptr));
+        }
+    }
+
+    if (supportedKeySystems.contains(GStreamerEMEUtilities::s_WidevineKeySystem) || supportedKeySystems.contains(GStreamerEMEUtilities::s_ClearKeyKeySystem)) {
+        for (const auto& mediaType : GStreamerEMEUtilities::s_webmEncryptionMediaTypes) {
+            gst_caps_append_structure(caps.get(), gst_structure_new_empty(mediaType.characters()));
+            gst_caps_append_structure(caps.get(), gst_structure_new("application/x-webm-enc", "original-media-type", G_TYPE_STRING, mediaType.characters(), nullptr));
+        }
+    }
+
+    return caps;
+}
+
+static void webkitMediaThunderParserConstructed(GObject* object)
+{
+    G_OBJECT_CLASS(webkit_media_thunder_parser_parent_class)->constructed(object);
+
+    auto self = WEBKIT_MEDIA_THUNDER_PARSER(object);
+
+    self->priv->decryptor = gst_element_factory_make("webkitthunder", nullptr);
+    self->priv->parser = makeGStreamerElement("parsebin"_s, "inner-parser"_s);
+
+    gst_bin_add_many(GST_BIN_CAST(self), self->priv->decryptor.get(), self->priv->parser.get(), nullptr);
+    gst_element_link(self->priv->decryptor.get(), self->priv->parser.get());
+
+    g_signal_connect(self->priv->parser.get(), "autoplug-factories", G_CALLBACK(+[](GstElement*, GstPad*, GstCaps* caps, gpointer) -> GValueArray* {
+        ALLOW_DEPRECATED_DECLARATIONS_BEGIN;
+        GValueArray* result;
+
+        GUniquePtr<GList> factories(gst_element_factory_list_get_elements(GST_ELEMENT_FACTORY_TYPE_DECODABLE, GST_RANK_MARGINAL));
```

Comment:
I think you have to call `gst_plugin_feature_list_free` or you'd be leaking the factories, according to doc.

A more complicated but maybe nicer would be tweaking the implementation of GUniquePtr to be able to make a quick inline declaration of the deleter.